docs/proposals/resource-quota-reservations.md

```diff
@@ -0,0 +1,228 @@
+T5. ADMIT
+
+RESTStorage
+T6. Create Object
+```
+
+**QUOTA_CONTROLLER**
+
+```
+T1. QUOTA := ...
+T2. OBSERVED_USAGE := ... [live queries based on latest data]
+T3. UPDATE QUOTA.STATUS.USED IFF OBSERVED USAGE != QUOTA.STATUS.USED
+```
+
+If the `QUOTA_CONTROLLER.T1` happens after `API_SERVER.T4`,
+and `QUOTA_CONTROLLER.T3` happens before `API_SERVER.T6` completes,
+then when the quota controller recalculates usage, it could have calculated
+usage without observing the outcome of the create.
+
+In practice, this scenario is *extremely rare*, but could happen if
+there was a network hiccup of some kind that blocked creation of
+the object at time `API_SERVER.T6` in etcd without causing the request
+to timeout.  The API server would have also been responding and reading
+from that same etcd to support a full quota recalculation and status
+update at time `QUOTA_CONTROLLER.T3`.
+
+The next controller pass would recalculate usage correctly.
+
+A solution should be put in place to mitigate this risk and a long-term
+plan to handle should be agreed upon.
+
+## Use Cases
+
+* As a `cluster-admin`, I want more reliable quota enforcement.
+
+### ResourceQuota reservations
+
+The recommended solution is for the quota status to track
+observed usage and reserved usage as separate values.
+
+The quota admission controller allows a request to be admitted
+if the `status.used + status.reserved + request <= status.hard`.
+If it chooses to admit the object, it will add a new reservation
+to the quota status that captures the `APIVersion`, `Kind`,
+and `UID` of the object that made the reservation.  It will also
+add a `ExpirationTime` that dictates how long that reservation
+is valid.
+
+The quota controller must be updated to become reservation aware
+in its synchronization loop.  It would calculate observed usage
+as before, but it would compare the set of observed resources
+against the set of current reservations.  If an object is observed,
+any corresponding reservation made by that object will be removed
+since its consumption will properly be tracked in usage.  If a
+reservation has expired, it will be removed from the quota document.
+A default expirationTime could be applied to sufficiently mitigate
+risk based on the anticipated quorum window across masters.
+
+In order to implement this solution, an admission controller must
+have access to the UID that will be assigned to an object at creation
+time.  This is currently not the case since UID assignment happens
+after admission control, and any UID that is assigned by admission
+control is overwritten.  The API server code should be updated
+to populate ObjectMetaSystemFields prior to invoking admission control.
+For this proposal, UID is required since Name would be insufficient
+to track against.  For a number of other use cases, code is simplified
+if Name is set as well (error-messages come to mind).
+
+## Data Model Impact
+
+```
+type ResourceQuotaReservation struct {
```

Comment:
@bgrant0607 - i know its been awhile, but what did you mean by hidden state?  is the transient reservation concept hidden state in your view?